Trabalho2/grafo.c

```diff
@@ -10,57 +10,412 @@
+            return -1;
+        }
+    }
 
+    ligacao **aux = origem->ligacoes;
+    origem->ligacoes = (ligacao**) realloc(origem->ligacoes, sizeof(ligacao*)*(origem->tamanho+1));
+
+    if(origem->ligacoes == NULL){
+        origem->ligacoes = aux; ////Garante que a memoria do origem->ligacoes não é perdida mesmo que o realloc não tenha tido sucesso
+        return -1;
+    }
+    
+    ligacao *lig = cria_lig(destino, peso);
+
+    if(lig == NULL){
+        return -1;
+    }
+
+    origem->ligacoes[origem->tamanho] = lig;
+    origem->tamanho++;
+
+    return 0;
 }
 
 
 no_grafo * encontra_no(grafo *g, char *nomeU)
 {
+    
+    if(g == NULL){
         return NULL;
+    }
+
+    if(nomeU == NULL){
+        return NULL;
+    }
+
+    for(int i=0;i<g->tamanho;i++)
+    {
+        if(strcmp(g->nos[i]->nome_user, nomeU) == 0){
+            return g->nos[i];
+        }
+    }
+
+    return NULL;
 }
 
 
 
 
 grafo * criaGrafo(tabela_dispersao *td)
 {
-    return NULL;
+    if(td == NULL){
+        return NULL;
+    }
+
+    grafo* g = grafo_novo();
+
+    if(g == NULL){
+        return NULL;
+    }
+
+    elemento *elem;
+    no_grafo *origem, *dest;
+    int msg[2];
+    msg[0] = msg[1] = 0;
+
+    for(int i=0;i<td->tamanho;i++){
+        elem = td->elementos[i];
+        while(elem != NULL)
+        {   
+            origem = no_insere(g, elem->msg->remetente);
+            dest = no_insere(g, elem->msg->destinatario);
+            
+            ligacao2(td, elem->msg->remetente, elem->msg->destinatario, msg);
+
+            if(msg[0]>0){
+                if(origem == NULL){
+                    origem = encontra_no(g, elem->msg->remetente);
+                }
+                if(dest == NULL){
+                    dest = encontra_no(g, elem->msg->destinatario);
+                }
+                cria_ligacao(origem, dest, msg[0]);
+            }
+
+            if(msg[1]>0){
+                if(origem == NULL){
+                    origem = encontra_no(g, elem->msg->remetente);
+                }
+                if(dest == NULL){
+                    dest = encontra_no(g, elem->msg->destinatario);
+                }
+                cria_ligacao(dest, origem, msg[1]);
+            }
+
+            elem = elem->proximo;
+        }
+    }
+
+    return g;
 }
 
 
 
 no_grafo **lista_amigos(grafo *g, char *nomeU, int *n)
 {
-   return NULL;
+    if(g == NULL){
+        return NULL;
+    }
+
+    if(nomeU == NULL){
+        return NULL;
+    }
+
+    if(n == NULL){
+        return NULL;
+    }
+
+    no_grafo ** amigos = NULL, **amigo =  NULL;
+    no_grafo *no;
+    int amigo1 = 0, count = 0;
+    *n = 0;
+
+    if((no = encontra_no(g, nomeU)) == NULL){
+        return NULL;
+    }
+
+    for(int i=0;i<no->tamanho;i++)
+    {
+        if(no->ligacoes[i]->peso_ligacao >= 4){
+            amigo1 = 1;
+            amigo = (no_grafo **) realloc(amigo, sizeof(no_grafo)*(count+1));

```

Comment:
you are not checking the allocation success

Reply:
Done